### ISSUE TITLE
parameter type of `any`

### ISSUE THREAD

**Issue by dcbaker** [cyrusfirheir/twee3-language-tools#56]
Right now a parameter has to have a distinct type, but I have a case where I have a macro that takes any for one parameter, does internal type checking, and then hands off to my backend. Having an `any`, that twee3-language-tools simply ignores so long as it is set would be very useful.

**Comment by cyrusfirheir** [comment]
You can set `"skipArgs": true` for that. Adding an `any` type would be sorta difficult. Given how varied parameter types are, it'd be a challenge to figure out where an `any` type ends and the next parameter starts.

**Comment by dcbaker** [comment]
Well, in my case I have a parameter of `bareword &+ bareword &+ any`, so I really don't want to skipArgs, as two of the arguments do have clear types.

Maybe I'm missing something, but paramters to sugarcube macros are a space delimited list of arguments, something roughly like: `<<macroName (\s+(!\s)*>>` (assuming a sort of psuedo regex were `!\s` means any non-whitespace character), or is the syntax more complicated than that?

**Comment by cyrusfirheir** [comment]
Links... The link markup throws a wrench into it all. It can have spaces in it.

Maybe I am missing something here too, but here's what I see as the problem:

If you want the `any` to disable checking for one parameter, where do you decide where that parameter ends?

**Comment by MinusGix** [comment]
> Links... The link markup throws a wrench into it all. It can have spaces in it.
> 
> Maybe I am missing something here too, but here's what I see as the problem:
> 
> If you want the `any` to disable checking for one parameter, where do you decide where that parameter ends?

As Cyrus says, the links are the more notable roadblock to simply splitting it by spaces. Then there is also expressions and strings.  
The problem of not knowing where `any` ends depends on how we define it. There's two methods:  
 - Make it special. As it is used in the above example as one parameter type and it just matches everything remaining. This is feasible to implement, but harder and could result in weird bugs.
 - Make it a normal type that is just any `ArgType`, (so if you want repeated data then you do `...any` like everything else). This can kinda already be emulated by doing a `text|passage|link|image` or so, since that spans over most arg types. I could add this easily.  
  
The downside of the second is that it doesn't act like a complete any type. A complete any type would essentially act like 'parse the first parameter type', 'parse the second parameter type', 'everything else gets used directly'. The second would be more accurately called `anyarg`.  
So, for a true any, it would be a special-type that then essentially skips over all args after it it activated. This is possible with the code, though as noted before, I worry about odd bugs that need to be squashed due to that. Still, I think I could do it.  
There is one issue though, is that we'd still be doing argument parsing for this (based on SugarCube's for accuracy) before parameter parsing. Thus, there would need to be some modifications to the argument parser for a mode that ignores errors (like `<<thing "hello>>` is an error if `skipArgs` is false, but fine if it is true, but we still want at least some types for those first types).  
So I think this is feasible, just might take a bit.

**Comment by dcbaker** [comment]
Ahh, I forgot you can use link syntax inside a `<<link>>` macro. Yeah, that does make for a rather ambiguous grammar. You'd almost have to parse that recursively, since as son as you run into `[[` you need to consume everything until the next `]] then parse that as a subexpression.

For my case it would be sufficient to have an `anyword`, that would essentially be a union of `barewrod`, `string`, `number`, and `boolean,` (in fact what I have in my config at the moment is basically that), which seems like it might be easier to implement as it would follow the same parsing rules as bareword, but allow numbers and booleans?

**Comment by MinusGix** [comment]
`text` covers that, I believe. It is a `bareword` or a `boolean` or a `number` or a `string` (or a `null` or a `NaN`).
`bareword &+ bareword &+ text`
Or do you mean that it would only allow a single word? (So the string `"alpha"` is allowed but `"alpha gamma"` is not)

**Comment by dcbaker** [comment]
no, I think `text` covers it. I thought that text was `bareword | string`. Any might still be useful, but for the case I had in mind `text` is suffcient :)

**Comment by MinusGix** [comment]
Glad that works. If you have any other issues/suggestions/questions, feel free to open issues :) 
I'll still be adding `any` because I think it would be useful (and would allow types for some macros we can't implement at the moment), I probably would have already finished with it but had a good amount of work over the last few days.
